[AutoDiscovery] Fix scheduling of configs

juliogreff opened the pull request:
### What does this PR do?

Fixes the scheduling of configs under the following conditions:

* When running `agent check`. Configs would not be scheduled in time, so `agent check` would not see them. Now `container` and `kubelet` config providers have a synchronous first `Collect`.
* When running `DD_LOGS_CONFIG_CCA_IN_AD`, a container that belongs to an AD-annotated pod would briefly have a `container_collect_all` config before honoring the one in the annotations.

### Describe how to test/QA your changes

* With both `DD_LOGS_CONFIG_CCA_IN_AD` set to `true` and `false`:
    * Check that `agent check` picks up AD-annotated pods
    * Check that the agent does not hang when starting up in a non-containerized environment
* With `DD_LOGS_CONFIG_CCA_IN_AD=true`, check that a pod with a logs annotation does not get scheduled, unscheduled, re-scheduled. It should get only scheduled once.

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

juliogreff committed:
[ad] Wait for first workloadmeta bundle on provider start

This fixes an issue in `agent check`, where a subtle change in timing
introduced in #12968 meant that configs did not get scheduled in time
for `agent check` to see them, effectively appearing as if they were not
present. This change ensures that `kubelet` and `container` providers
only return from `Collect` if they've received at least one event bundle
from workloadmeta.

This also changes workloadmeta to actually allow the sending of empty
bundles, as that's necessary to signal that the subscription has started
and there are indeed no entities in the store. Otherwise, config
providers in non-containerized environments will hang indefinitely.

juliogreff committed:
[ad] Generate configs only once for k8s containers

Previously, we'd generate configs for the container events themselves,
and again when handling the pod event. When the container belonged to an
AD-annotated pod, it would briefly have a `container_collect_all` config
before honoring the one in the annotations. Now, since the configs are
generated entirely at once, AutoDiscovery can ignore the
`container_collect_all` correctly the first time.

juliogreff committed:
[ad] Add container_collect_all after annotation configs

djmitche commented on pkg/autodiscovery/providers/container.go:
```diff
@@ -73,7 +75,7 @@ func (d *ContainerConfigProvider) Collect(ctx context.Context) ([]integration.Co
 }
 
-func (d *ContainerConfigProvider) listen() {
+func (d *ContainerConfigProvider) listen(ch chan struct{}) {
```
```suggestion
// listen, closing the given channel after the initial set of events are received
func (d *ContainerConfigProvider) listen(ranOnceCh chan struct{}) {
```
(just for clarity of intention)

djmitche commented on pkg/autodiscovery/providers/kubecontainer.go:
```diff
@@ -163,19 +164,21 @@ func (k *KubeContainerConfigProvider) generateConfig(e workloadmeta.Entity) ([]i
-			configs[idx].Source = names.Container + ":" + containerEntityName
+		// kubernetes containers need to be handled together with their
+		// pod, so they generate a single []integration.Config.
+		// otherwise, with container_collect_all, it's possible for a
+		// container that belongs to an AD-annotated pod to briefly
+		// have a container_collect_all when it shouldn't.
```
```suggestion
		// otherwise, it's possible for a container that belongs
		// to an AD-annotated pod to briefly be scheduled without its
		// annotations.
```

djmitche reviewed (approved):
Some comments on the comments ;)

juliogreff committed:
[ad] Clarify code comments

juliogreff committed:
[ad] Send empty changesets from KubeContainerConfigProvider

The config poller expects a changeset to be able to mark a streaming
provider as started. Otherwise, `AutoConfig.LoadAndRun` will block,
potentially forever, breaking AD.

juliogreff committed:
[ad] Preserve order in generated configs

This is a follow up to an earlier commit in this PR, that attempted to
generate container_collect_all configs after all others. That commit
missed the mangling done by `processEvents`. Now it does preserve order
correctly.
